XenModel/XenAPI/HTTP.cs

```diff
@@ -150,45 +163,90 @@ private static string ReadLine(Stream stream)
         /// </summary>
         /// <param name="stream"></param>
         /// <returns>True if a redirect has occurred - headers will need to be resent.</returns>
-        private static bool ReadHttpHeaders(ref Stream stream, IWebProxy proxy, bool nodelay, int timeout_ms)
+        private static bool ReadHttpHeaders(ref Stream stream, IWebProxy proxy, bool nodelay, int timeout_ms, List<string> headers = null)
         {
-            string response = ReadLine(stream);
-            int code = getResultCode(response);
+            // read headers/fields
+            string line = ReadLine(stream), initialLine = line, transferEncodingField = null;
+            if (string.IsNullOrEmpty(initialLine)) // sanity check
+                return false;
+            while (!string.IsNullOrEmpty(line) && !string.IsNullOrWhiteSpace(line))

```

Comment:
IsNullOrWhiteSpace checks for empty string too, so the first check is not necessary (you could add it as a reminder/comment at the end of the line because the method name doesn't sound like it).